confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -18,133 +18,258 @@
-                if 200 <= response.status_code <= 299:
-                    return response.json(cls=decoder, **kwargs)
-                raise SchemaRegistryClientError(response.status_code,
-                                                response.json().get('message'))
-            except ValueError:
-                raise SchemaRegistryClientError(response.status_code,
-                                                response.content)
+        if body is not None:
+            body = json.dumps(body, cls=self.JSONDuckEncoder)
+            _headers = {'Content-Length': str(len(body)),
+                        'Content-Type': "application/vnd.schemaregistry.v1+json"}
 
-    def register_schema(self, subject_name, schema):
+        response = self.session.request(
+            method, url="/".join([self.base_url, url]),
+            headers=_headers, data=body)
+
+        try:
+            if 200 <= response.status_code <= 299:
+                return response.json()
+            raise SchemaRegistryError(response.json().get('message', None),
+                                      response.json().get('error_code', None))
+        # Schema Registry may return HTML when it hits unexpected errors
+        except ValueError as e:
+            print(str(e))
+            raise SchemaRegistryError(response.status_code,
+                                      response.content)
+
+    class JSONDuckEncoder(json.JSONEncoder):
+        def default(self, obj):
+            """
+            Enables JSON encoding for any class with a to_json method.
+
+            Args:
+                - obj (object): object implementing ``to_json(self)``
+
+            Returns:
+                str: JSON encoded obj
+
+            """
+            return obj.to_json()
+
+
+class _SchemaCache(object):
+
+    def __init__(self):
+        self.lock = Lock()
+
+        self.schema_id_index = {}
+        self.schema_index = {}
+
+    def add_schema(self, schema_id, schema):
+        """
+        Add a Schema identified by schema_id to the cache.
+
+        Args:
+            schema_id (int): Schema's registration id
+            schema (Schema): Schema instance
+
+        Returns:
+            int: The schema_id
+        """
+        # Don't overwrite existing keys
+        schema = self.get_schema(schema_id)
+        if schema is not None:
+            return
+
+        with self.lock:
+            # Pessimistic
+            if schema_id not in self.schema_id_index:
+                self.schema_id_index[schema_id] = schema
+
+    def get_schema(self, schema_id):
+        """
+        Get the schema instance associated  with schema_id from the cache.
+
+        Args:
+            schema_id (int): Id used to identify a schema
+        Returns:
+            Schema: The schema if known; else None
+
+        """
+        return self.schema_id_index.get(schema_id, None)
+
+    def add_schema_id(self, schema, schema_id):
+        """
+        Add this Schema's id to the cache.
```

Comment:
Is there a reason to not combine both schema2id and id2schema in the same add_..()?